File: easy/704-binary-search.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = int(0)
        right = len(nums) - 1

        while left <= right:
            middle = (left + right) // 2
            if nums[middle] < target:
                left = middle + 1
            elif nums[middle] > target:
                right = middle - 1
            else:
                return middle
        return -1

print(Solution().search([-1,0,3,5,9,12], 9))

File: easy/704-binary-search.cxx
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int start = 0;
        int end = nums.size() - 1;
        int mid = (start + end) / 2;
        
        while (start <= end) {
            if (nums[mid] > target) {
                end = mid - 1;
            } else if (nums[mid] < target) {
                start = mid + 1;
            } else {
                break;
            }
            mid = (start + end) / 2;
        }
        
        return nums[mid] == target ? mid: -1;
    }
};